File: hello-swift/Sources/Server/Protocol.swift
import Foundation

/// Command category.
enum CommandType {
    case stream, data, query
}

/// Stream lifecycle commands (client → server).
enum StreamCommand: String, CaseIterable {
    case CREATE, COMPLETE, CLOSE
}

/// Data transfer commands (client → server).
enum DataCommand: String, CaseIterable {
    case READ
}

/// Query commands (client → server).
enum QueryCommand: String, CaseIterable {
    case GET_STATUS, LIST_STREAMS
}

/// Parsed command descriptor.
struct CommandInfo {
    let cmdType: CommandType
    let streamCmd: StreamCommand?
    let dataCmd: DataCommand?
    let queryCmd: QueryCommand?
}

/// Lookup table: command string → CommandInfo
let commandLookup: [String: CommandInfo] = {
    var map: [String: CommandInfo] = [:]
    for c in StreamCommand.allCases {
        map[c.rawValue] = CommandInfo(cmdType: .stream, streamCmd: c, dataCmd: nil, queryCmd: nil)
    }
    for c in DataCommand.allCases {
        map[c.rawValue] = CommandInfo(cmdType: .data, streamCmd: nil, dataCmd: c, queryCmd: nil)
    }
    for c in QueryCommand.allCases {
        map[c.rawValue] = CommandInfo(cmdType: .query, streamCmd: nil, dataCmd: nil, queryCmd: c)
    }
    return map
}()

File: hello-go/src/server/protocol.go
package server

// CommandType classifies a protocol command.
type CommandType int

const (
	CommandTypeStream CommandType = iota
	CommandTypeData
	CommandTypeQuery
)

// StreamCommand is a stream lifecycle command.
type StreamCommand int

const (
	CommandCreate   StreamCommand = iota
	CommandComplete StreamCommand = iota
	CommandClose    StreamCommand = iota
)

// DataCommand is a data-operation command.
type DataCommand int

const (
	CommandRead DataCommand = iota
)

// QueryCommand is a query command.
type QueryCommand int

const (
	CommandGetStatus   QueryCommand = iota
	CommandListStreams QueryCommand = iota
)

// CommandInfo is the result of parsing a command string.
type CommandInfo struct {
	cmdType   CommandType
	streamCmd StreamCommand
	dataCmd   DataCommand
	queryCmd  QueryCommand
}

func (c CommandInfo) isStream() bool { return c.cmdType == CommandTypeStream }
func (c CommandInfo) isData() bool   { return c.cmdType == CommandTypeData }
func (c CommandInfo) isQuery() bool  { return c.cmdType == CommandTypeQuery }
